crates/wasi/src/preview2/preview1/mod.rs

```diff
@@ -253,143 +743,599 @@ impl<
     async fn fd_pwrite<'a>(
         &mut self,
         fd: types::Fd,
         ciovs: &types::CiovecArray<'a>,
         offset: types::Filesize,
     ) -> Result<types::Size, types::Error> {
-        todo!()
+        let desc = self.descriptors().await?.get_descriptor(fd).cloned();
+        let n = match desc {
+            Some(Descriptor::File(File { fd, blocking, .. })) if self.table().is_file(fd) => {
+                let Some(buf) = first_non_empty_ciovec(ciovs)? else {
+                    return Ok(0)
+                };
+                let stream = self
+                    .write_via_stream(fd, offset)
+                    .await
+                    .context("failed to call `write-via-stream`")
+                    .map_err(types::Error::trap)?;
+                if blocking {
+                    self.blocking_write(stream, buf)
+                } else {
+                    wasi::streams::Host::write(self, stream, buf)
+                }
+                .await
+                .map_err(|_| types::Errno::Io)?
+            }
+            Some(Descriptor::Stdout(..) | Descriptor::Stderr(..)) => {
+                // NOTE: legacy implementation returns SPIPE here
+                return Err(types::Errno::Spipe.into());
+            }
+            _ => return Err(types::Errno::Badf.into()),
+        }
+        .try_into()
+        .or(Err(types::Errno::Overflow))?;
+        Ok(n)
     }
 
+    /// Return a description of the given preopened file descriptor.
+    #[instrument(skip(self))]
     async fn fd_prestat_get(&mut self, fd: types::Fd) -> Result<types::Prestat, types::Error> {
-        todo!()
+        if let Some(Descriptor::PreopenDirectory((_, p))) =
+            self.descriptors().await?.get_descriptor(fd)
+        {
+            let pr_name_len = p.len().try_into().or(Err(types::Errno::Overflow))?;
+            return Ok(types::Prestat::Dir(types::PrestatDir { pr_name_len }));
+        }
+        Err(types::Errno::Badf.into()) // NOTE: legacy implementation returns BADF here
     }
 
+    /// Return a description of the given preopened file descriptor.
+    #[instrument(skip(self))]
     async fn fd_prestat_dir_name<'a>(
         &mut self,
         fd: types::Fd,
         path: &GuestPtr<'a, u8>,
         path_max_len: types::Size,
     ) -> Result<(), types::Error> {
-        todo!()
+        let path_max_len = path_max_len.try_into().or(Err(types::Errno::Overflow))?;
+        if let Some(Descriptor::PreopenDirectory((_, p))) =
+            self.descriptors().await?.get_descriptor(fd)
+        {
+            // NOTE: This conditional is not valid, the bug is preserved from legacy implementation
+            // for backwards-compatibility
+            // https://github.com/bytecodealliance/wasmtime/blob/b9e4474f1f04ae15ca27ca0ed695d798e1c0a295/crates/wasi-common/src/snapshots/preview_1.rs#L507-L514
+            if p.len() < path_max_len {
+                return Err(types::Errno::Nametoolong.into());
+            }
```

Comment:
lets definitely not preserve that bug :). That one has been around for a while! Lets get rid of this check and follow up with a PR to fix the legacy impl, and a test. I expect that fix will only be around for its final release, but its still worth doing.